### PR TITLE
Add lookup-in function

### PR DESCRIPTION
**Pull request by fhitchen**
Java SDK 2.3 added support for sub-document functions to retrieve and modify parts of documents. I have added a lookup-in function that takes parameters of bucket, index and whatever document paths you would like to retrieve. I did this as I have a use case where I just need a single value from a large document and pulling the whole document for a single field is a waste of network resources. You can also check the existence of a path using this function. It will return a {:path nil} key value pair to indicate the absence of a JSON index. 

One interesting thing to note, originally I converted the JSON path's to keywords in the return map, for example :document[0].index.purpose, but found lookups gave a compilation error as [] are not valid in keywords. These now get replaced by <> which are valid,

There are a set of tests to exercise the lookup-in function using a new document.

The corresponding sub-document mutation has not been tackled. I may see if I have time to do it but I don't have any need for it at the moment.

If you accept these changes can you also publish a new version to clojars.org?

regards, Francis.

### CODE REVIEW

File: src/earthen/clj_cb/bucket.clj [hunk 1]
```diff
@@ -56,6 +56,19 @@
...
          (document->map doc))))))
 
+(defn lookup-in
+  "Retrieves sub-document values"
+  [bucket id & rest]
+  (let [builder (.lookupIn bucket id)
+        _ (.get builder (into-array rest))
+        result (.execute builder)]
+    (apply merge-with into {} (map-indexed
+                               (fn [idx itm]
```
Comment (Alfrheim) [inline src/earthen/clj_cb/bucket.clj]: What is idx and itm? index and item? don't be afraid of being more expressive, github don't charge extra for characters :)

File: src/earthen/clj_cb/bucket.clj [hunk 2]
```diff
@@ -56,6 +56,19 @@
...
+                                 {(keyword (clojure.string/replace itm #"\[|\]" {"[" "<" "]" ">"}))
+                                  (if (contains? (supers (class (.content result idx))) com.couchbase.client.java.document.json.JsonValue)
```
Comment (Alfrheim) [inline src/earthen/clj_cb/bucket.clj]: If I understood correctly, what we are trying to do here is something like:
               (if (is-json-value result idx)
                      (read-content-as-json result idx)
                      (.content result idx)......?

File: README.md [hunk 3]
```diff
@@ -27,22 +33,37 @@ A Clojure java-client wrapper for Couchbase Server 4.
...
         
-        (b/replace bucket "Aaron1" {:hitpoints 23832, :level 141, :loggedIn true, :name "Aaron1", :jsonType "player", :experience 248, :uuid "78edf902-7dd2-49a4-99b4-1c94ee286a33"})
-            => {:id "Aaron1", :cas 6654687454080, :expiry 0, :mutation-token nil, :content {:hitpoints 23832, :level 141, :loggedIn true, :name "Aaron1", :jsonType "player", :experience 248, :uuid "78edf902-7dd2-49a4-99b4-1c94ee286a33"}}
+        (b/replace! bucket "Aaron1" {:name "Aaron1", :hitpoints 23832, :experience 248, :level 141, :loggedIn true, :tues [{:symptom {:asthma "Ventolin Inhler"}} {:symptom {:pain "Asprin"}}], :sponsors ["Nike" "Reebok" "Freddies"], :uuid "78edf902-7dd2-49a4-99b4-1c94ee286a33", :jsonType "player"})
+
+        => {:id "Aaron1", :cas 1569768233881239552, :expiry 0, :mutation-token nil, :content {:name "Aaron1", :hitpoints 23832, :experience 248, :level 141, :loggedIn true, :tues [{:symptom {:asthma "Ventolin Inhler"}} {:symptom {:pain "Asprin"}}], :sponsors ["Nike" "Reebok" "Freddies"], :uuid "78edf902-7dd2-49a4-99b4-1c94ee286a33", :jsonType "player"}}
+
+       ;; To lookup particular objects / values in a document, pass a list of paths. Vectors in
```
Comment (Alfrheim) [inline README.md:55]: how about we put that comment also in the method as a documentation? 
I think it's worth it :+1:

File: README.md [hunk 4]
```diff
@@ -17,6 +17,12 @@ A Clojure java-client wrapper for Couchbase Server 4.
     
     ;; we create a bucket
     (def bucket (c/open-bucket cluster "gamesim-sample"))
+
+    ;; we authenticate with the bucket password, note: the password parameter on the c/bucket-open
```
Comment (Alfrheim) [inline README.md:21]: That comment will help to have it in the c/authenticate method too

**Review by Alfrheim — COMMENTED** [review]
Nice job! 
Let me know what do you think about the comments.

**Comment by Alfrheim** [comment]
I will create a new version once it is merged. Also I will update to clojars :)

File: src/earthen/clj_cb/bucket.clj (hunk 2 above)
Comment (fhitchen) [reply src/earthen/clj_cb/bucket.clj]: Correct, it is either one of these com.couchbase.client.java.document.json.JsonArray or JsonObject backed by the JsonValue object or a primitive type, string, integer.

File: src/earthen/clj_cb/bucket.clj (hunk 1 above)
Comment (fhitchen) [reply src/earthen/clj_cb/bucket.clj]: Sure, will make the change, was just copying from an example using map-indexed.

File: README.md (hunk 3 above)
Comment (fhitchen) [reply README.md:55]: Done.

File: README.md (hunk 4 above)
Comment (fhitchen) [reply README.md:21]: Done.

**Review by Alfrheim — APPROVED** [review]

**Comment by Alfrheim** [comment]
Just created new version https://clojars.org/earthen/clj-cb/versions/0.2.4-SNAPSHOT

**Pull request merged**